File: AOJ/vol5/0514.java
import java.util.*;

public class Main {
    public static void main(String[] args) throws Exception {

        Scanner scan = new Scanner(System.in);
        
        while(true){
            //parts number
            int supply = scan.nextInt();
            int motor = scan.nextInt();
            int cable = scan.nextInt();
            if (supply==0){
                break;
            }
            

            int len = scan.nextInt();
            int[] res = new int[supply+motor+cable];
			Arrays.fill(res, 2);
			ArrayList<Integer> intList = new ArrayList<Integer>();
            int len2 = 0;
            
            // pass 
            for (int i=0; i<len; i++){
                int no1 = scan.nextInt()-1;
                int no2 = scan.nextInt()-1;
                int no3 = scan.nextInt()-1;
                int ans = scan.nextInt();
                if (ans==1){
                    res[no1]=1;
                    res[no2]=1;
                    res[no3]=1;
                }else{
                    intList.add(no1);
                    intList.add(no2);
                    intList.add(no3);
                    len2+=3;
                }
            }
            
            //not pass
            for (int i=0; i<len2; i+=3){
                int no1 = intList.get(i);
                int no2 = intList.get(i+1);
                int no3 = intList.get(i+2);
                if (res[no1]==1 && res[no2]==1){
                    res[no3]=0;
                }
                else if (res[no3]==1 && res[no2]==1){
                    res[no1]=0;
                }
                else if (res[no1]==1 && res[no3]==1){
                    res[no2]=0;
                }
            }
            
            for (int i=0;i<supply+motor+cable; i++){
                System.out.println(res[i]);
            }
            
            
        }
    }
}

File: AOJ/vol5/0514.py
while True:
    supply,motor,cable=(int(x) for x in input().split())
    if supply==0 and motor==0:
        break
    checknum=int(input())
    
    reslist=[2]*(supply+motor+cable)
    stack=[]
    for i in range(checknum):
        no1,no2,no3,result=(int(x)-1 for x in input().split())
        result+=1
        if result==1:
            reslist[no1], reslist[no2], reslist[no3]=1,1,1
        else:
            stack.append((no1,no2,no3))

    for i in range(len(stack)):
        t=stack[i]
        no1,no2,no3=t[0],t[1],t[2]
        if reslist[no1]==1 and reslist[no2]==1:
            reslist[no3]=0
        elif reslist[no3]==1 and reslist[no2]==1:
            reslist[no1]=0
        elif reslist[no3]==1 and reslist[no1]==1:
            reslist[no2]=0
            
    for i in range(len(reslist)):
        print(reslist[i])